stan/math/rev/constraint/stochastic_column_constrain.hpp

```diff
@@ -26,53 +27,45 @@ namespace math {
 template <typename T, require_rev_matrix_t<T>* = nullptr>
 inline plain_type_t<T> stochastic_column_constrain(const T& y) {
   using ret_type = plain_type_t<T>;
-  const Eigen::Index N = y.rows();
-  const Eigen::Index M = y.cols();
-  using eigen_mat_rowmajor
-      = Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>;
-  arena_t<eigen_mat_rowmajor> x_val(N + 1, M);
+
+  const auto N = y.rows();
+  const auto M = y.cols();
+  arena_t<T> arena_y = y;
+
+  arena_t<ret_type> arena_x = stochastic_column_constrain(arena_y.val_op());
+
   if (unlikely(N == 0 || M == 0)) {
-    return ret_type(x_val);
-  }
-  arena_t<change_eigen_options_t<T, Eigen::RowMajor>> arena_y = y;
-  arena_t<eigen_mat_rowmajor> arena_z(N, M);
-  using arr_vec = Eigen::Array<double, 1, -1>;
-  arr_vec stick_len = arr_vec::Constant(M, 1.0);
-  for (Eigen::Index k = 0; k < N; ++k) {
-    const double log_N_minus_k = std::log(N - k);
-    arena_z.row(k)
-        = inv_logit(arena_y.array().row(k).val_op() - log_N_minus_k).matrix();
-    x_val.row(k) = stick_len.array() * arena_z.array().row(k);
-    stick_len -= x_val.array().row(k);
+    return arena_x;
   }
-  x_val.row(N) = stick_len;
-  arena_t<ret_type> arena_x = x_val;
-  reverse_pass_callback([arena_y, arena_x, arena_z]() mutable {
-    const Eigen::Index N = arena_y.rows();
-    auto arena_x_arr = arena_x.array();
-    auto arena_y_arr = arena_y.array();
-    auto arena_z_arr = arena_z.array();
-    auto stick_len_val = arena_x.array().row(N).val().eval();
-    auto stick_len_adj = arena_x.array().row(N).adj().eval();
-    for (Eigen::Index k = N; k-- > 0;) {
-      arena_x_arr.row(k).adj() -= stick_len_adj;
-      stick_len_val += arena_x_arr.row(k).val();
-      stick_len_adj += arena_x_arr.row(k).adj() * arena_z_arr.row(k);
-      auto arena_z_adj = arena_x_arr.row(k).adj() * stick_len_val;
-      arena_y_arr.row(k).adj()
-          += arena_z_adj * arena_z_arr.row(k) * (1.0 - arena_z_arr.row(k));
+
+  reverse_pass_callback([arena_y, arena_x]() mutable {
+    const auto M = arena_y.cols();
+
+    const auto& x_val = to_ref(arena_x.val_op());
+    const auto& x_adj = to_ref(arena_x.adj_op());
+
+    for (Eigen::Index i = 0; i < M; ++i) {
+      // backprop for softmax
+      Eigen::VectorXd x_pre_softmax_adj
+          = -x_val.col(i) * x_adj.col(i).dot(x_val.col(i))
+            + x_val.col(i).cwiseProduct(x_adj.col(i));
```

Comment:
You can just make the memory `x_pre_softmax_adj` once
```suggestion
    Eigen::VectorXd x_pre_softmax_adj(x_val.rows());
    for (Eigen::Index i = 0; i < M; ++i) {
      // backprop for softmax
      x_pre_softmax_adj.noalias()
          = -x_val.col(i) * x_adj.col(i).dot(x_val.col(i))
            + x_val.col(i).cwiseProduct(x_adj.col(i));
```

Reply:
You could also change `sum_to_zero_vector_backprop` to just take in a matrix